### PR TITLE
curvefs: support creating hybrid fs type

### PR DESCRIPTION
**Pull request by Cyber-SiKu**
### What problem does this PR solve?

Issue Number: #1356  

[4.1](https://github.com/opencurve/curve/blob/d5f8188fe51d60c38591827be7f71579ed27d3c6/docs/cn/curvefs-client%E7%BB%9F%E4%B8%80%E7%BC%93%E5%AD%98%E8%AE%BE%E8%AE%A1.md#41-%E5%88%9B%E5%BB%BA%E6%96%87%E4%BB%B6%E7%B3%BB%E7%BB%9F%E5%92%8C%E5%88%9D%E5%A7%8B%E5%8C%96%E6%B5%81%E7%A8%8B)

Problem Summary:

support hybrid fs creation

### What is changed and how it works?

What's Changed:

1. add hybrid fs type in proto
2. add create hybird fs in mds
3. add create hybird fs in curvefs/tool

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/tools/create/curvefs_create_fs.cpp [hunk 1]
```diff
@@ -65,22 +68,34 @@ void CreateFsTool::PrintHelp() {
...
+              << "]\n[-fsType=s3 -s3_ak=" << FLAGS_s3_ak
               << " -s3_sk=" << FLAGS_s3_sk
               << " -s3_endpoint=" << FLAGS_s3_endpoint
               << " -s3_bucket_name=" << FLAGS_s3_bucket_name
               << " -s3_blocksize=" << FLAGS_s3_blocksize
               << " -s3_chunksize=" << FLAGS_s3_chunksize
-              << "] [-mdsAddr=" << FLAGS_mdsAddr
-              << "] [-rpcTimeoutMs=" << FLAGS_rpcTimeoutMs
-              << "-rpcRetryTimes=" << FLAGS_rpcRetryTimes << "]" << std::endl;
+              << "]\n[-fsType=hybird -volumeSize=" << FLAGS_volumeSize
```
Comment (h0hmj) [inline curvefs/src/tools/create/curvefs_create_fs.cpp:87]: same typo

File: curvefs/src/tools/create/curvefs_create_fs.cpp [hunk 2]
```diff
@@ -170,10 +188,31 @@ int CreateFsTool::Init() {
...
         volume->set_bitmaplocation(location);
         request.mutable_fsdetail()->set_allocated_volume(volume);
-    } else {
-        std::cerr << "-fsType should be " << kFsTypeS3 << " or "
-                  << kFsTypeVolume << "." << std::endl;
-        ret = -1;
+        return 0;
+    };
+
+    auto SetHybirdRequest = [&]() -> int {
```
Comment (h0hmj) [inline curvefs/src/tools/create/curvefs_create_fs.cpp:194]: same typo

File: curvefs/src/tools/create/curvefs_create_fs.cpp [hunk 3]
```diff
@@ -170,10 +188,31 @@ int CreateFsTool::Init() {
...
+        if (SetS3Request() != 0 || SetVolumeRequest() != 0) {
+            return -1;
+        }
+        request.set_fstype(common::FSType::TYPE_HYBRID);
+        return 0;
+    };
+
+    std::unordered_map<std::string, std::function<int()>> setRequestFuncMap{
+        {kFsTypeS3, SetS3Request}, {kFsTypeVolume, SetVolumeRequest},
+        {kFsTypeHybird, SetHybirdRequest}
```
Comment (h0hmj) [inline curvefs/src/tools/create/curvefs_create_fs.cpp:204]: same typo

File: curvefs/src/tools/create/curvefs_create_fs.cpp [hunk 4]
```diff
@@ -170,10 +188,31 @@ int CreateFsTool::Init() {
...
+    };
+
+    auto func = setRequestFuncMap.find(FLAGS_fsType);
+
+    if (func == setRequestFuncMap.end()) {
+        std::cerr << "fsType should be one of [S3, Volume, Hybird]";
```
Comment (h0hmj) [inline curvefs/src/tools/create/curvefs_create_fs.cpp:210]: same typo

File: curvefs/src/tools/curvefs_tool_define.h [hunk 5]
```diff
@@ -136,6 +136,7 @@ const char kEtcdFollowerValue[] = "StateFollower";
 /* fs type */
 const char kFsTypeS3[] = "s3";
 const char kFsTypeVolume[] = "volume";
+const char kFsTypeHybird[] = "hybird";
```
Comment (h0hmj) [inline curvefs/src/tools/curvefs_tool_define.h:139]: same typo

**Comment by h0hmj** [comment]
change your commit message
"curvefs/add_hybird_fs_before" is confusing and typo error

should be something like:
curvefs: add TYPE_HYBRID fs creation support

**Comment by aspirer** [comment]
> change your commit message "curvefs/add_hybird_fs_before" is confusing and typo error
> 
> should be something like: curvefs: add TYPE_HYBRID fs creation support

I have changed.....

**Comment by h0hmj** [comment]
> 

commit message not pr message

**Comment by h0hmj** [comment]
i am not sure whether this proto change will break other function or not, need more developers' confirm

File: curvefs/src/mds/mds_service.cpp [hunk 6]
```diff
@@ -36,76 +36,107 @@ void MdsServiceImpl::CreateFs(::google::protobuf::RpcController* controller,
...
+            const auto& volume = request->fsdetail().volume();
+            LOG(INFO) << "CreateFs request, fsName = " << fsName
+                      << ", blockSize = " << blockSize
+                      << ", volume.volumeName = " << volume.volumename()
+                      << ", enableSumInDir = " << enableSumInDir
+                      << ", owner = " << request->owner()
+                      << ", capacity = " << request->capacity();
+
+            FSStatusCode status =
+                fsManager_->CreateFs(request, response->mutable_fsinfo());
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply curvefs/src/mds/mds_service.cpp:62]: > when createfs, will finally call FsInfoWrapper(const ::curvefs::mds::CreateFsRequest* request, uint64_t fsId, uint64_t rootInodeId) which doesn't use fstype info from request directly, you need fix it.

agree, related code :https://github.com/opencurve/curve/blob/7aa7f5d6d335c0178edcfe1589b8933235b0fb3a/curvefs/src/mds/fs_info_wrapper.h#L72

File: curvefs/src/tools/create/curvefs_create_fs.cpp [hunk 7]
```diff
@@ -170,10 +188,31 @@ int CreateFsTool::Init() {
...
         request.mutable_fsdetail()->set_allocated_volume(volume);
-    } else {
-        std::cerr << "-fsType should be " << kFsTypeS3 << " or "
-                  << kFsTypeVolume << "." << std::endl;
-        ret = -1;
+        return 0;
+    };
+
+    auto SetHybirdRequest = [&]() -> int {
+        if (SetS3Request() != 0 || SetVolumeRequest() != 0) {
```
[the remark this answers is not in the document]
Comment (h0hmj) [reply curvefs/src/tools/create/curvefs_create_fs.cpp:195]: > What's the logic mean? hybrid is one of s3 or volume?

hybrid need both s3 and volume info, so it can r/w two backends

File: curvefs/src/mds/mds_service.cpp [hunk 8]
```diff
@@ -36,76 +36,107 @@ void MdsServiceImpl::CreateFs(::google::protobuf::RpcController* controller,
...
+                           << ", s3Info.bucketname = " << s3Info.bucketname()
+                           << ", enableSumInDir = " << enableSumInDir
+                           << ", owner = " << request->owner()
+                           << ", capacity = " << request->capacity()
+                           << ", errCode = " << FSStatusCode_Name(status);
+                return;
+            }
+        };
+
+    auto createHybirdFs = [&]() {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/mds/mds_service.cpp:112]: > Hybrid

fix

File: curvefs/src/mds/mds_service.cpp [hunk 9]
```diff
@@ -36,76 +36,107 @@ void MdsServiceImpl::CreateFs(::google::protobuf::RpcController* controller,
...
+
+    switch (type) {
+        case ::curvefs::common::FSType::TYPE_VOLUME:
+            createVolumeFs();
+            break;
+        case ::curvefs::common::FSType::TYPE_S3:
+            createS3Fs();
+            break;
+        case ::curvefs::common::FSType::TYPE_HYBRID:
+            createHybirdFs();
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/mds/mds_service.cpp:125]: > same typo

fix

File: curvefs/src/tools/create/curvefs_create_fs.cpp (hunk 1 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/create/curvefs_create_fs.cpp:87]: > same typo

fix

File: curvefs/src/tools/create/curvefs_create_fs.cpp (hunk 2 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/create/curvefs_create_fs.cpp:194]: > same typo

fix

File: curvefs/src/tools/create/curvefs_create_fs.cpp (hunk 3 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/create/curvefs_create_fs.cpp:204]: > same typo

fix

File: curvefs/src/tools/create/curvefs_create_fs.cpp (hunk 4 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/create/curvefs_create_fs.cpp:210]: > same typo

fix

File: curvefs/src/tools/curvefs_tool_define.h (hunk 5 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_define.h:139]: > same typo

fix

File: curvefs/src/mds/fs_info_wrapper.h [hunk 10]
```diff
@@ -69,13 +69,23 @@ class FsInfoWrapper {
...
-            fsInfo.set_fstype(FSType::TYPE_S3);
-            fsInfo.set_capacity(request->capacity());
-        } else {
-            fsInfo.set_fstype(FSType::TYPE_VOLUME);
-            fsInfo.set_capacity(
-                std::min(detail.volume().volumesize(), request->capacity()));
+
+        switch (request->fstype()) {
+            case FSType::TYPE_VOLUME:
+                fsInfo.set_fstype(FSType::TYPE_S3);
```
Comment (cw123) [inline curvefs/src/mds/fs_info_wrapper.h:75]: case volume , set s3?

File: curvefs/src/mds/fs_info_wrapper.h [hunk 11]
```diff
@@ -69,13 +69,23 @@ class FsInfoWrapper {
...
+                fsInfo.set_capacity(request->capacity());
+                break;
+            case FSType::TYPE_S3:
+                fsInfo.set_fstype(FSType::TYPE_VOLUME);
```
Comment (cw123) [inline curvefs/src/mds/fs_info_wrapper.h:79]: type is wrong

File: curvefs/src/mds/fs_info_wrapper.h [hunk 12]
```diff
@@ -69,13 +69,23 @@ class FsInfoWrapper {
...
+                break;
+            case FSType::TYPE_VOLUME:
+                fsInfo.set_fstype(FSType::TYPE_VOLUME);
+                fsInfo.set_capacity(std::min(detail.volume().volumesize(),
+                                             request->capacity()));
+                break;
+            case FSType::TYPE_HYBRID:
+                fsInfo.set_fstype(FSType::TYPE_HYBRID);
+                // TODO(huyao): set capacity for hybrid fs
+                fsInfo.set_capacity(std::min(detail.volume().volumesize(),
```
Comment (h0hmj) [inline curvefs/src/mds/fs_info_wrapper.h:86]: do we have any request validation before calling this FsInfoWrapper()? such like if type=hybrid, but volume or s3info is missing in fsdetail. they are optional.

File: curvefs/src/tools/create/curvefs_create_fs.cpp [hunk 13]
```diff
@@ -65,22 +68,34 @@ void CreateFsTool::PrintHelp() {
               << "] [-capacity=" << FLAGS_capacity
               << "] [-blockSize=" << FLAGS_blockSize
               << "] [-enableSumInDir=" << FLAGS_enableSumInDir
-              << "] [-fsType=volume -volumeSize=" << FLAGS_volumeSize
+              << "] [-mdsAddr=" << FLAGS_mdsAddr
```
Comment (h0hmj) [inline curvefs/src/tools/create/curvefs_create_fs.cpp:71]: mdsAddr and other flags? seems not related to this pr?

**Review by SeanHai — APPROVED** [review]

**Review by h0hmj — APPROVED** [review]
